### PR TITLE
[SPARK-7402][ML] JSON SerDe for standard param types

### PR DESCRIPTION
**Pull request by mengxr**
This PR implements the JSON SerDe for the following param types: `Boolean`, `Int`, `Long`, `Float`, `Double`, `String`, `Array[Int]`, `Array[Double]`, and `Array[String]`. The implementation of `Float`, `Double`, and `Array[Double]` are specialized to handle `NaN` and `Inf`s. This will be used in pipeline persistence. @jkbradley

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/param/params.scala [hunk 1]
```diff
@@ -287,6 +422,16 @@ class StringArrayParam(parent: Params, name: String, doc: String, isValid: Array
 
   /** Creates a param pair with a [[java.util.List]] of values (for Java and Python). */
   def w(value: java.util.List[String]): ParamPair[Array[String]] = w(value.asScala.toArray)
+
+  override def jsonEncode(value: Array[String]): String = {
+    import org.json4s.JsonDSL._
```
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/ml/param/params.scala]: This is not imported globally because it shadows `++`.

**Comment by jkbradley** [comment]
Making a pass now

File: mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala [hunk 2]
```diff
@@ -21,6 +21,120 @@ import org.apache.spark.SparkFunSuite
...
+    { // BooleanParam
+      val param = new BooleanParam(dummy, "name", "doc")
+      for (value <- Seq(true, false)) {
+        val json = param.jsonEncode(value)
+        assert(param.jsonDecode(json) === value)
+      }
+    }
+
+    { // IntParam
+    val param = new IntParam(dummy, "name", "doc")
```
Comment (jkbradley) [inline mllib/src/test/scala/org/apache/spark/ml/param/ParamsSuite.scala]: Indentation here and in a few places below.

**Comment by jkbradley** [comment]
Nice, I can't really see anything to correct.  LGTM except for the indentation nit.

**Comment by mengxr** [comment]
test this please

**Comment by mengxr** [comment]
Merged into master.

**Pull request closed without merging**